java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java

```diff
@@ -21,24 +21,37 @@
 
 import org.apache.arrow.vector.IntVector;
 import org.apache.arrow.vector.complex.impl.IntWriterImpl;
-import org.apache.arrow.vector.complex.writer.IntWriter;
 import org.apache.avro.io.Decoder;
 
 /**
  * Consumer which consume int type values from avro decoder.
  * Write the data to {@link IntVector}.
  */
-public class AvroIntConsumer implements Consumer {
+public class AvroIntConsumer extends Consumer {
 
-  private final IntWriter writer;
+  private final IntWriterImpl writer;
 
+  /**
+   * Instantiate a AvroIntConsumer.
+   */
   public AvroIntConsumer(IntVector vector) {
     this.writer = new IntWriterImpl(vector);
+    this.nullable = vector.getField().isNullable();
+    if (nullable) {
+      getNullFieldIndex(vector.getField());
+    }
   }
 
   @Override
   public void consume(Decoder decoder) throws IOException {
-    writer.writeInt(decoder.readInt());
+    if (!nullable) {
```

Comment:
did you consider using decomposition?  Something like:
```
class NullableTypeConsumer {
   Consumer delegate;
   int nullIndex = 0;
   @Override 
   public void consume(Decoder decoder) {
       if (nullIndex != decoder.readInt()) {
           delegate.consume(decoder);
       } else {
          writer.setPosition(writer.getPosition() + 1)
       }
      
   }
}
```

Reply:
Good suggestion, this will be really helpful in Union complex types. 👍